### PR TITLE
derp: unify server's clientSet interface into concrete type

### PR DESCRIPTION
**Pull request by bradfitz**
73280595a8880bdca for #2751 added a "clientSet" interface to
distinguish the two cases of a client being singly connected (the
common case) vs tolerating multiple connections from the client at
once. At the time (three years ago) it was kinda an experiment
and we didn't know whether it'd stop the reconnect floods we saw
from certain clients. It did.

So this promotes it to a be first-class thing a bit, removing the
interface. The old tests from 73280595a were invaluable in ensuring
correctness while writing this change (they failed a bunch).

But the real motivation for this change is that it'll permit a future
optimization to add flow tracking for stats & performance where we
don't contend on Server.mu for each packet sent via DERP. Instead,
each client can track its active flows and hold on to a *clientSet and
ask the clientSet per packet what the active client is via one atomic
load rather than a mutex. And if the atomic load returns nil, we'll
know we need to ask the server to see if they died and reconnected and
got a new clientSet. But that's all coming later.

Updates #3560

### CODE REVIEW

File: derp/derp_server.go [hunk 1]
```diff
@@ -194,26 +192,45 @@ type Server struct {
...
 }
 
-// singleClient is a clientSet of a single connection.
-// This is the common case.
-type singleClient struct{ c *sclient }
+func (s *clientSet) Len() int {
+	if s.dup != nil {
+		return len(s.dup.set)
+	}
+	return 1
```
Comment (maisem) [inline derp/derp_server.go:217]: should this check if `activeClient` is not nil?

File: derp/derp_server.go [hunk 2]
```diff
@@ -194,26 +192,45 @@ type Server struct {
...
+	// activeClient holds the currently active connection for the set. It's nil
+	// if there are no connections or the connection is disabled.
+	//
+	// A pointer to a clientSet can be held by peers for long periods of time
+	// without holding Server.mu to avoid mutex contention on Server.mu, only
+	// re-acquiring the mutex and checking the clients map if activeClient is
+	// nil.
+	activeClient atomic.Pointer[sclient]
+
+	// dup is non-nil if there are multiple connections for the
```
Comment (maisem) [inline derp/derp_server.go:205]: say that it contains all clients?

File: derp/derp_server.go [hunk 3]
```diff
@@ -194,26 +192,45 @@ type Server struct {
...
+// ForeachClient calls f for each client in the set.
+//
+// The Server.mu must be held.
+func (s *clientSet) ForeachClient(f func(*sclient)) {
+	if s.dup != nil {
+		for c := range s.dup.set {
+			f(c)
+		}
+	} else {
+		if c := s.activeClient.Load(); c != nil {
```
Comment (maisem) [inline derp/derp_server.go:229]: `else if` ?

File: derp/derp_server.go [hunk 4]
```diff
@@ -595,38 +616,57 @@ func (s *Server) unregisterClient(c *sclient) {
...
+	set, ok := s.clients[c.key]
+	if !ok {
 		c.logf("[unexpected]; clients map is empty")
-	case singleClient:
+		return
+	}
+
+	dup := set.dup
+	if dup == nil {
+		cur := set.activeClient.Load()
```
Comment (maisem) [inline derp/derp_server.go:627]: `// common case` ?

File: derp/derp_server.go (hunk 2 above)
Comment (maisem) [reply derp/derp_server.go:205]: add a few more docs describing when `activeClient` or `dup` are (nil, empty), (nil, non-empty), (non-nil, non-empty), (non-nil, empty)

**Review by maisem — APPROVED** [review]

File: derp/derp_server.go (hunk 1 above)
Comment (bradfitz) [reply derp/derp_server.go:217]: Actually I'm going to try to remove Len entirely, as it's barely used (mostly for logging) and not very well defined.

Comment (bradfitz) [reply derp/derp_server.go:217]: I'll make it return 0 on nil, but turns out that only affects some log statements that should never be hit.

I'll remove Len later probably, in a separate change.

**Review by raggi — APPROVED** [review]
I think the disablement issue, if it's there, is a preexisting one, so this change is LGTM, but I think we should make the disablement behavior more clear, and reduce the overlap in state to avoid bugs / hard behavior assertions. Similarly I think sendHistory is a bit of a misnomer right now, so maybe we can cleanup the whole dup state a fair bit.

File: derp/derp_server.go [hunk 5]
```diff
@@ -239,17 +261,15 @@ type dupClientSet struct {
...
+	if d == nil {
+		return s.activeClient.Load()
 	}
-	return nil
-}
-func (s *dupClientSet) Len() int { return len(s.set) }
-func (s *dupClientSet) ForeachClient(f func(*sclient)) {
-	for c := range s.set {
-		f(c)
+	if d.last != nil && !d.last.isDisabled.Load() {
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply derp/derp_server.go:269]: Honestly I forget how it all works and I haven't touched it in 3 years so it's all out of my head. I tried to make the minimal diff and port over existing behavior unchanged (which fortunately was easier by strict tests keeping me honest)

But yeah, I want to simplify/document it in another change once I understand it again.

**Pull request merged**